### ISSUE TITLE
reducing complexity, then test coverage via cyclomatic complexity

### ISSUE THREAD

**Issue by BMaxV** [panda3d/panda3d#1491]
## Description
So, I found this tool that I like

https://pypi.org/project/radon/

It measures complexity. It can help us identify and prioritize pain points prioritize them.

It counts the "paths" through a function, and every "if" increases the number of paths by 1. 

I wrote a little crawler that looks at a project/repo/tree and makes it convenient to look at the thing as a whole.

The idea would be to encapsulate small parts of problematic functions. We could then write tests for those now small and easy to understand functions.

## Use Case

# how to

A being good, F being the worst and very bad, all of panda3d currently looks like this:

![piechart](https://user-images.githubusercontent.com/5713211/233784796-0215e972-3bc3-4296-9335-338bc0bf7791.svg)

And it helps navigating to e.g. `/direct/src/actor/Actor.py` which looks like this:

![radonpieActor](https://user-images.githubusercontent.com/5713211/233784894-e5286602-a2bd-4a98-8290-d0b48d8dc4e8.svg)

```
Actor.py
    M 1773:4 Actor.getAnimControls - E
    M 130:4 Actor.__init__ - D
    M 1899:4 Actor.loadModel - C
    M 2144:4 Actor.loadAnims - C
```

getAnimControls

https://github.com/panda3d/panda3d/blob/master/direct/src/actor/Actor.py#L1773

This block could be wrapped into a function

`actor.getAnimControlDictItems(lodName) -> partDict, animControlDictItems`

```
if lodName is None or self.mergeLODBundles:
    # Get all LOD's
    animControlDictItems = self.__animControlDict.items()
else:
    partDict = self.__animControlDict.get(lodName)
    if partDict is None:
        Actor.notify.warning("couldn't find lod: %s" % (lodName))
        animControlDictItems = []
    else:
        animControlDictItems = [(lodName, partDict)]
```

I think this would:
 * create small, easy to review pull requests. 
 * good first issues, since it's not required to understand how panda words, all that's necessary is making sure all required variables are passed to the function and all assigned variables are returned.
 * writing tests should be easier if the function that needs to be tested is small

I would really like to get in the habit of contributing small improvements like this. So it would be cool if we could figure something out to make that happen @Moguri @rdb

**Comment by rdb** [comment]
I appreciate the suggestion, but I worry that this would just create a big review burden without meaningfully improving the code quality.  I think Godhart's law applies here.  The goal in refactoring should never be to make some number go up, but to create a meaningful improvement to maintainability that's worth the risk of touching working code.

Because the risk of breaking working code goes down substantially with increased test coverage, I don't really want to entertain PRs that refactor code that has no high-quality unit tests, at least not if there is no other compelling benefit.

I would rather focus on creating those tests first, which creates an opportunity for people to contribute easy-to-review pull requests.

**Comment by BMaxV** [comment]
Yes, that's why I put it up as a draft for now. It should only be done when all the parts and the full function as it is right now is fully tested.

> The goal in refactoring should never be to make some number go up, but to create a meaningful improvement to maintainability that's worth the risk of touching working code.

The goal is to create subparts of code that are easier to test.

E.g. it is or should be lot easier to test that something like `actor.getAnimControlDictItems(lodName) -> partDict, animControlDictItems` behaves as expected, rather than testing the complex behavior of the "parent" function.

Do we have any tests for actor right now? `/tests/actor/` looks empty to me, but I'm not sure if that's the place where I should look for them.

I'm also not sure what this feature is and how to use it effectively, so it's a good incentive to create the appropriate documentation and samples. It get it's about LOD somehow?
